README.md
![](https://github.com/Abhilash-Chandran/vega_embed_flutter/workflows/build%20and%20test/badge.svg)
[![pub package](https://img.shields.io/pub/v/vega_embed_flutter.svg)](https://pub.dev/packages/vega_embed_flutter)

# vega_embed_flutter

A Flutter widget to embed vega-lite charts to flutter web and flutter apps using webview_flutter. Works consistently in Flutter web as of now. Internally it uses vega-embed js library to embed these specs using JS interops and `HTMLElementView`.

Vega-lite charts are visual grammar based charts which are specified using vega-lite specifications in the form of json files. 

A Demo page with examples for different vega-embed options is available below.

https://abhilash-chandran.github.io/vega_embed_flutter/

[![A demo page](repo_example.png)](https://abhilash-chandran.github.io/vega_embed_flutter/)

A full fledged demo page with various charts utilizing this package is available [here](https://abhilash-chandran.github.io/vega_embed_flutter_demo_page/#/)

[![Demo Page Gif](interactions.gif)](https://abhilash-chandran.github.io/vega_embed_flutter_demo_page/#/)

A webview_flutter based widget is also exposed which allows rendering vega-lite charts in mobile devices.
This feature is still experimental. Please report any issues in this regard.
![Flutter web view demo](web_view_demo.png)
# Credits

This widget is made possible only because of the great works done by the team behind [vega-embed](https://github.com/vega/vega-embed), [vega-Lite](https://vega.github.io/vega-lite/) and [vega](https://vega.github.io/vega/). For a detailed documentation of how visualize using Vega-lite specs please head to their [doumentation](https://vega.github.io/vega-lite/docs/) page.

# General Setup Steps for flutter web

## Import Vega related Javascript files.

Start by adding the `script` tag for vega related java script files. For example.

**Index.html**

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <script src="https://cdn.jsdelivr.net/npm/vega@5"></script>
    <script src="https://cdn.jsdelivr.net/npm/vega-lite@4"></script>
    <script src="https://cdn.jsdelivr.net/npm/vega-embed@6"></script>
    <title>vega_flutter</title>
  </head>
  <body>
    <script src="main.dart.js" type="application/javascript"></script>
  </body>
</html>
```

## Add your vega lite specification files

Vega-Lite and Vega defines its visualization in the form of json file. So for you project it would be easy to maintain all this vega-lite spec files in a folder say _vega_lite_specs_ folder. For example

```
$Project_root\vega_lite_specs
              |_ bart_chart.json
              |_ interactive_multiline_plot.json
              |_ interactive_splom.json
```

## Add assets entry in pupspec.yaml

In order for the web build to pick up the json schemas add it your `pubspec.yaml`. For e.g.

```
flutter:
  uses-material-design: true
  # This line includes all the files these directories during the build process
  # and placess them under build/web/assets folder.
  assets:
    - vega_lite_specs/
```

## Import the file and use the widget

Just import the `vega_embed_flutter` library as below

`import 'package:vega_embed_flutter/vega_embed_flutter.dart';`

Create a normal stateless/ful widget and use it as a normal stateless/ful widget.

```
class BarChart extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VegaLiteEmbedder(
      viewFactoryId: 'MyBarChart',
      vegaLiteSpecLocation: '/assets/vega_lite_specs/bar_chart.json',
    );
  }
}
```

# General set for normal flutter app

Follow the [Add your vega lite schema files](https://pub.dev/packages/vega_embed_flutter#add-your-vega-lite-schema-files) section and [Add assets entry in pupspec.yaml](https://pub.dev/packages/vega_embed_flutter#add-assets-entry-in-pupspecyaml) as mentioned above.

## Import web view specific file and use the widget

Import the vega embed view specific library as shown below
`import 'package:vega_embed_flutter/vega_embed_webview.dart';`

Create a normal stateless/ful widget and use it as a normal stateless/ful widget.

```
class ExampleWebViewApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: SafeArea(
        child: Center(
          child: Column(
            children: [
              Expanded(
                child: VegaLiteWebViewEmbedder(
                  vegaLiteSpecLocation: 'vega_lite_specs/bar_chart copy.json',
                ),
              ),
              Expanded(
                child: VegaLiteWebViewEmbedder(
                  vegaLiteSpecLocation:
                      'vega_lite_specs/interactive_index_chart.vl.json',
                  vegaEmbedOptions: VegaEmbedOptions(theme: 'dark'),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

# VegaEmbed Options [WIP](available only for web)

Following vega-embed options are tested and are available as part of demo page. If you need further comprehensive options to be tested or enabled please do raise and issue so that I can spend time on options which are necessary.

For a detailed description of what each vega embed options check this page.
https://github.com/vega/vega-embed#options

|  Embed Options   |      Working       |                                                                                                  Comments                                                                                                  |
| :--------------: | :----------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
|       mode       |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |
|      theme       |       :100:        | Check [here](https://github.com/vega/vega-themes#included-themes) for detials on the themes. Valid values are [excel, ggplot2, quartz, vox, fivethirtyeight , dark, latimes, urbaninstitute, googlecharts] |
|   defaultStyle   | :heavy_check_mark: |                                                                                                                                                                                                            |
|     renderer     | :heavy_check_mark: |                                                                                                                                                                                                            |
|     logLevel     |   :construction:   |                                                                                                                                                                                                            |
|     tooltip      |     :thinking:     |                                                                                         Let me know if you need it                                                                                         |
|      loader      |        :x:         |                                                                                                                                                                                                            |
|      patch       |        :x:         |                                                                                                                                                                                                            |
|      width       | :heavy_check_mark: |                                                                          Note: :warning: width mentioned in spec overrides this.                                                                           |
|      height      | :heavy_check_mark: |                                                                          Note: :warning: Height mentioned in spec overrides this.                                                                          |
|     padding      | :heavy_check_mark: |                                                                                                                                                                                                            |
|     actions      |   :construction:   |                                                            Note: :warning: an example is shown to remove actions. But its not complete to spec.                                                            |
|   scaleFactor    | :heavy_check_mark: |                                                                     Note: This is a scale factor for the exported file, not the plot.                                                                      |
|      config      |     :thinking:     |                                                         Let me know if you need it. Porting config object to dart will take lot of effort to test.                                                         |
|    editorUrl     |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |
|   sourceHeader   |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |
|   sourceFooter   |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |
|      hover       |     :thinking:     |                                                                                        Let me know if you need it.                                                                                         |
|       i18n       | :heavy_check_mark: |                                                                                                                                                                                                            |
| downloadFileName | :heavy_check_mark: |                                                                                                                                                                                                            |
|   formatLocale   |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |
| timeFormatLocale |   :construction:   |                                                                                          Avialble but not tested.                                                                                          |

#### Any kind of code, idea, suggestions are welcome. Please check the repository and file your feedbacks and bugs which is highly appreciated.


CHANGELOG.md
## [0.7.0-null-safety] 12 May 2021

- Migrated code base to support null safety.

## [0.6.2-0.6.3] 27 August 2020

- Some fixes for improving the scores.
- Remove menu from the webview render of vega charts when no actions provided.
- add relevant style sheet for vega-embed in webview.

## [0.6.0+3] 24 June 2020

- Upgraded freezed package version.
- Edited the example demo page.

## [0.6.0] - [0.6.0+2] 11 June 2020

- Added support for webview to enable vega lite chart for mobile.
- utilizes the flutter webview_flutter

## [0.5.0] - [0.5.0+2] - 23 April 2020.

- Added more vega embed options and tested.
- Added more examples for embed options.
- Now the demo page includes code samples.
- Edited the example md to include live page and corrected some spelling errors.

## [0.1.1+1] - [0.1.1+3] - 17 March 2020.

- edited the example section.
- added git hub actions.

## [0.1.1] - 16 March 2020.

- Now the widget accepts and passes the VegaEmbedOptions.
- Created a full fledged demo page.
- Example includes better documentation.

## [0.1.0] - 10 March 2020.

- Removed double Rendering.
  - Now uses the vega-embed and vega-tool tip styles directly.
- Multiple Charts in Single page is working now.
- Changed the example folder structure.
- Renamed VegaLiteSchemLocation to VegaLiteSpecLocation.
- Differentiating Schema vs Spec.

## [0.0.1] - 06 March 2020.

- A basic version of vega_embed for flutter.
  - Works with single charts.
  - Interactions works as well.

TODO:

- [X] Enable flutter webview to support mobile platform.
- Extend vega-embed options to fully support all options.
- [x] ~~Provide more vega-lite references.~~
- [x] ~~Document the theming section~~
- [x] ~~Add a section vega-embed options.~~
- [x] ~~Add a definitive and documented demo. A basic example is available in the example folder.~~
- [x] ~~Set up a live page with multitude of charts.~~
- [x] ~~Enable vega-embed theming.~~
- [x] ~~Work on rendering multiple charts in the same page.~~


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:vega_embed_flutter/vega_embed_flutter.dart';
import 'examples/examples.dart';
import 'code_box.dart';

void main() => runApp(RootApp());

class RootApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return ExampleApp();
  }
}

class ExampleApp extends StatefulWidget {
  @override
  _ExampleAppState createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp>
    with SingleTickerProviderStateMixin {
  int selectedIndex = 0;
  late TabController _controller;

  @override
  void initState() {
    super.initState();
    _controller = TabController(length: 2, vsync: this);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Vega embed demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                Container(
                  decoration: selectedIndex != 0
                      ? null
                      : BoxDecoration(
                          border: Border(
                            bottom: BorderSide(
                              color: Colors.deepOrange,
                              width: 2.0,
                            ),
                          ),
                        ),
                  // padding: EdgeInsets.symmetric(horizontal: 6.0),
                  child: TextButton(
                    onPressed: () {
                      setState(() {
                        selectedIndex = 0;
                        _controller.animateTo(0);
                      });
                    },
                    style: TextButton.styleFrom(
                      backgroundColor: selectedIndex == 0
                          ? Colors.orangeAccent
                          : Colors.amber[100],
                    ),
                    child: Text(
                      'Plots Demo',
                      style: TextStyle(backgroundColor: Colors.transparent),
                    ),
                  ),
                ),
                Container(
                  decoration: selectedIndex != 1
                      ? null
                      : BoxDecoration(
                          border: Border(
                            bottom: BorderSide(
                              color: Colors.deepOrange,
                              width: 2.0,
                            ),
                          ),
                        ),
                  padding: EdgeInsets.symmetric(horizontal: 6.0),
                  child: TextButton(
                    style: TextButton.styleFrom(
                      backgroundColor: selectedIndex == 1
                          ? Colors.orangeAccent
                          : Colors.amber[100],
                    ),
                    onPressed: () {
                      setState(() {
                        selectedIndex = 1;
                        _controller.animateTo(1);
                      });
                    },
                    child: Text(
                      'Embed Options Demo',
                      style: TextStyle(backgroundColor: Colors.transparent),
                    ),
                  ),
                ),
              ],
            ),
            Divider(),
            Expanded(
              flex: 10,
              child: TabBarView(
                controller: _controller,
                children: [
                  MainDemoPage(),
                  PlotWithOptions(),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class MainDemoPage extends StatelessWidget {
  const MainDemoPage({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Row(children: [
      Expanded(
        flex: 2,
        child: Column(
          children: <Widget>[
            Expanded(
              child: Container(
                decoration: BoxDecoration(
                    border: Border(
                  right: BorderSide(color: Colors.purple),
                  bottom: BorderSide(color: Colors.purple),
                )),
                child: InteractiveMultiLineChart(),
              ),
            ),
            Expanded(
              child: Container(
                decoration: BoxDecoration(
                    border: Border(
                  right: BorderSide(color: Colors.purple),
                )),
                child: BarChart(),
              ),
            ),
          ],
        ),
      ),
      Expanded(
        flex: 3,
        child: InteractiveSplom(),
      ),
    ]);
  }
}

/// A simple bar chart with inline data.
class BarChart extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VegaLiteEmbedder(
      viewFactoryId: 'MyBarChart',
      vegaLiteSpecLocation: 'assets/vega_lite_specs/bar_chart.json',
    );
  }
}

/// An interactive multi line plot with hover tool tip.
/// Uses the VegaEmbedOptions to set the theme as 'quartz'.
class InteractiveSplom extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VegaLiteEmbedder(
      viewFactoryId: 'MyInteractiveSplom',
      vegaLiteSpecLocation: 'assets/vega_lite_specs/interactive_splom.json',
      vegaOptions: VegaEmbedOptions(
        theme: 'latimes',
      ),
    );
  }
}

/// An interactive scatter plot of the Iris data set.
/// Uses the VegaEmbedOptions to set the theme as 'dark'.
class InteractiveMultiLineChart extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VegaLiteEmbedder(
      viewFactoryId: 'InteractiveMultiLine',
      vegaLiteSpecLocation:
          'assets/vega_lite_specs/interactive_multiline_plot.json',
      vegaOptions: VegaEmbedOptions(
        theme: 'dark',
      ),
    );
  }
}

class PlotWithOptions extends StatefulWidget {
  @override
  _PlotWithOptionsState createState() => _PlotWithOptionsState();
}

class _PlotWithOptionsState extends State<PlotWithOptions> {
  final PageController _pageController = PageController();
  final StreamController<List<int>> _pageIndexStreamCntlr =
      StreamController<List<int>>.broadcast();
  late Stream<List<int>> pageIndexStream;
  @override
  void initState() {
    super.initState();
    pageIndexStream = _pageIndexStreamCntlr.stream;
    Future.delayed(
        Duration(
          milliseconds: 16,
        ), () {
      _pageIndexStreamCntlr.add([1, 14, 14]);
    });
  }

  @override
  void dispose() {
    _pageController.dispose();
    pageIndexStream.drain();
    _pageIndexStreamCntlr.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    var bundle = DefaultAssetBundle.of(context);
    return Row(
      children: [
        Expanded(
          flex: 2,
          child: Container(
            padding: EdgeInsets.only(top: 4.0),
            child: Material(
              elevation: 8.0,
              child: ListView(
                children: [
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    0,
                    'Dark Theme',
                    hlStart: 14,
                    hlEnd: 14,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    1,
                    'Setting different theme',
                    hlStart: 13,
                    hlEnd: 13,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    2,
                    'Adjust height and width',
                    hlStart: 14,
                    hlEnd: 16,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    3,
                    'Render as svg',
                    hlStart: 15,
                    hlEnd: 15,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    4,
                    'Always open menu',
                    hlStart: 15,
                    hlEnd: 15,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    5,
                    'Custom Menu style',
                    hlStart: 15,
                    hlEnd: 20,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    6,
                    'Add padding to plot',
                    hlStart: 14,
                    hlEnd: 17,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    7,
                    'Change Menu Language',
                    hlStart: 14,
                    hlEnd: 17,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    8,
                    'Change Scale of export',
                    hlStart: 13,
                    hlEnd: 14,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    9,
                    'Remove Editor and Source options from menu',
                    hlStart: 13,
                    hlEnd: 16,
                  ),
                  ListButton(
                    _pageController,
                    _pageIndexStreamCntlr,
                    10,
                    'Customize Export file name',
                    hlStart: 13,
                    hlEnd: 13,
                  ),
                ],
              ),
            ),
          ),
        ),
        Expanded(
          flex: 5,
          child: StreamBuilder<List<int>>(
            stream: _pageIndexStreamCntlr.stream,
            builder: (context, streamSnap) {
              if (streamSnap.hasData) {
                bundle = DefaultAssetBundle.of(context);
                var currIndex = streamSnap.data![0];
                var file = 'lib/examples/example${currIndex}.dart';
                return FutureBuilder<String>(
                    future: bundle.loadString(file, cache: true),
                    builder: (context, snapshot) {
                      if (snapshot.hasData) {
                        return CodeBox(
                          code: snapshot.data ?? '',
                          viewId: 'codeBox_${currIndex}',
                          key: ValueKey('codeBox_${currIndex}'),
                          hlStart: streamSnap.data![1],
                          hlEnd: streamSnap.data![2],
                        );
                      }
                      return Center(
                        child: CircularProgressIndicator(
                          backgroundColor: Colors.deepOrange,
                        ),
                      );
                    });
              }
              return Center(
                child: CircularProgressIndicator(
                  backgroundColor: Colors.black,
                ),
              );
            },
          ),
        ),
        Expanded(
            flex: 7, child: PlotOptionsView(pageController: _pageController)),
      ],
    );
  }
}

class ListButton extends StatelessWidget {
  final PageController controller;
  final StreamController<List<int>> _pageIndexController;
  final int index;
  final String label;
  final int hlStart;
  final int hlEnd;
  const ListButton(
      this.controller, this._pageIndexController, this.index, this.label,
      {this.hlStart = 0, this.hlEnd = 0, key})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    Widget _child = InkWell(
      onTap: () {
        controller.jumpToPage(index);
        _pageIndexController.add([index + 1, hlStart, hlEnd]);
        // _pageIndexController.add(index + 1);
      },
      child: ListTile(
        title: Text(label),
      ),
    );
    return StreamBuilder<List<int>>(
        stream: _pageIndexController.stream,
        builder: (context, snapshot) {
          if (snapshot.hasData && snapshot.data![0] == index + 1) {
            return Ink(
              color: Colors.orangeAccent,
              child: _child,
            );
          }
          return _child;
        });
  }
}

class PlotOptionsView extends StatefulWidget {
  final PageController pageController;
  PlotOptionsView({required this.pageController});
  @override
  _PlotOptionsViewState createState() => _PlotOptionsViewState();
}

class _PlotOptionsViewState extends State<PlotOptionsView> {
  @override
  Widget build(BuildContext context) {
    return PageView(
      controller: widget.pageController,
      children: [
        Example1(),
        Example2(),
        Example3(),
        Example4(),
        Example5(),
        Example6(),
        Example7(),
        Example8(),
        Example9(),
        Example10(),
        Example11(),
      ],
    );
  }
}
